File: ScanbotSDKDemo/GenericDocumentsDemo/GenericDocumentResultViewController.m
//
//  GenericDocumentResultViewController.m
//  SBSDK Internal Demo
//
//  Created by Sebastian Husche on 22.05.20.
//  Copyright © 2020 doo GmbH. All rights reserved.
//

#import "GenericDocumentResultViewController.h"
#import "GenericDocumentFieldTableViewCell.h"

@interface GenericDocumentResultViewController () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic, strong) SBSDKGenericDocument *document;
@property (nonatomic, strong) NSArray<SBSDKGenericDocument *> *flatDocument;
@property (nonatomic, strong) UIImage* sourceImage;
@property (nonatomic, strong) IBOutlet UITableView *tableView;

@end

@implementation GenericDocumentResultViewController

+ (instancetype)makeWithResult:(SBSDKGenericDocument *)document sourceImage:(UIImage *)sourceImage {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    GenericDocumentResultViewController *controller =
    [storyboard instantiateViewControllerWithIdentifier:@"GenericDocumentResultViewController"];
    controller.document = document;
    controller.sourceImage = sourceImage;
    return controller;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.flatDocument = [self.document flatDocumentIncludingEmptyChildren:NO includingEmptyFields:NO];
    
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight = 100.0f;
    [self.tableView reloadData];
}

- (IBAction)shareSourceImage:(id)sender {
    if (self.sourceImage == nil) {
        return;
    }
    
    UIActivityViewController *activityViewController
    = [[UIActivityViewController alloc]initWithActivityItems:@[self.sourceImage] applicationActivities:nil];
    activityViewController.popoverPresentationController.sourceView = self.view;
    [self presentViewController:activityViewController animated:YES completion:nil];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.flatDocument.count;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return self.flatDocument[section].type.displayText;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.flatDocument[section].fields.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    GenericDocumentFieldTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"fieldCell"];
    SBSDKGenericDocumentField *field = self.flatDocument[indexPath.section].fields[indexPath.row];
    cell.fieldTypeLabel.text = field.type.name;
    
    UIImage *image = [field.image sbsdk_limitedToSize:CGSizeMake(10000.0, 80.0f)];
    cell.fieldImageView.image = image;
    
    if (field.value.text.length > 0) {
        cell.recognizedTextLabel.text = field.value.text;

        cell.recognizedTextInfoLabel.text
        = [NSString stringWithFormat:@"Recognized text with confidence %0.2f %%", field.value.confidence * 100.0f];
        
    } else {
        cell.recognizedTextInfoLabel.text = @"";
        cell.recognizedTextLabel.text = @"";
    }
    return cell;
}

- (NSString *)textForType:(SBSDKGenericDocumentType *)type {
    if (type.displayText != nil) {
        return type.displayText;
    }
    return type.name;
}

@end

File: ScanbotSDKDemo/Identity Detectors/GenericDocumentsDemo/GenericDocumentResultViewController.swift
//
//  GenericDocumentResultViewController.swift
//  ScanbotSDKDemo
//
//  Created by Danil Voitenko on 23.11.21.
//  Copyright © 2021 doo GmbH. All rights reserved.
//

import UIKit
import ScanbotSDK

final class GenericDocumentResultViewController: UIViewController {
    var document: SBSDKGenericDocument?
    var flatDocument: [SBSDKGenericDocument]?
    var sourceImage: UIImage?
    @IBOutlet var tableView: UITableView?
    
    static func make(with document: SBSDKGenericDocument, sourceImage: UIImage) -> GenericDocumentResultViewController {
        let storyboard = UIStoryboard(name: "Main", bundle: nil)
        let controller = storyboard.instantiateViewController(withIdentifier: "GenericDocumentResultViewController") as!
        GenericDocumentResultViewController
        controller.document = document
        controller.sourceImage = sourceImage
        return controller
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        flatDocument = document?.flatDocument(includeEmptyChildren: false, includeEmptyFields: false)
        tableView?.rowHeight = UITableView.automaticDimension
        tableView?.estimatedRowHeight = 100.0
        tableView?.reloadData()
    }
    
    @IBAction private func shareSourceImage(_ sender: Any) {
        guard let sourceImage = sourceImage else { return }
        
        let activityViewController = UIActivityViewController(activityItems: [sourceImage]
                                                              , applicationActivities: nil)
        activityViewController.popoverPresentationController?.sourceView = view
        present(activityViewController, animated: true, completion: nil)
    }
    
    private func text(for type: SBSDKGenericDocumentType) -> String {
        if let displayText = type.displayText {
            return displayText
        }
        return type.name
    }
}

extension GenericDocumentResultViewController: UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return flatDocument?.count ?? 0
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return flatDocument?[section].type.displayText
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return flatDocument?[section].fields.count ?? 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "fieldCell", for: indexPath) as!
        GenericDocumentFieldTableViewCell
        let field = flatDocument?[indexPath.section].fields[indexPath.row]
        cell.fieldTypeLabel.text = field?.type.name
        
        let image = field?.image?.sbsdk_limited(to: CGSize(width: 10000.0, height: 80.0))
        cell.fieldImageView.image = image
        
        if let value = field?.value, value.text.isEmpty == false {
            cell.recognizedTextLabel.text = value.text
            cell.recognizedTextInfoLabel.text = "Recognized text with confidence\(value.confidence * 100)"
        } else {
            cell.recognizedTextLabel.text = ""
            cell.recognizedTextInfoLabel.text = ""
        }
        return cell
    }
}
